ops/framework.py

```diff
@@ -138,8 +138,11 @@ def __repr__(self):
     def defer(self):
         """Defer the event to the future.
 
-        After being deferred, this event will be re-emited the next time there's
-        any other event for the charm.
+        After being deferred, this event will be re-emitted the next time there's
+        any other event for the charm. However the current event handler will
+        continue execution, in the absence of an explicit return statement. Further
+        when the event is re-emitted, the corresponding event handler will being
+        execution again from its first executable statement.
```

Comment:
I wrote the thing below, but I'd like to pester @jameinel because I'm sure there are better (clearer, as well as more correct and readable) ways of saying what I'm trying to say here.

```suggestion
        Deferring an event puts it and its handler(s) into a queue, and the next
        time the charm is invoked (by either an action or an event other than
        collect-metrics) the queue of events are dispatched again to the named
        event handlers.

        From the above you may deduce, but it's important to point out that:

        * ``defer()`` does not interrupt the execution of the current event
          handler. In almost all cases a call to ``defer()`` should be followed
          by an explicit ``return`` from the handler;

        * the re-execution of the deferred event handler starts from the top of
          the handler method, and

        * any deferred events get processed before the event (or action) that
          caused the current invocation of the charm.

        You shouldn't ``defer()`` an event if the same event at a later date is
        the thing that will unblock the progress of the handler. That is, for
        example, a ``config-changed`` event handler need not call ``defer()`` in
        wait of a following ``config-changed`` event to unblock its progress; it
        should instead just return (setting status appropriately perhaps).

        In general, there is little cause to ``defer()`` an event because the
        situation that would cause the event to be unblocked could normally be
        detected by the event that caused the event to get re-emitted and could
        be handled by it directly, as well as not necessarily be detected by the
        re-emitted event that gets fired before the current one. That is,

        1. event A fired, event A handler detects situation X, defers.

        2. event B fired, event A re-emitted, event A handler detects situation
           still X, defers again; event B handler resolves situation X.

        3. (some time later) event Z fired, event A re-emitted, event A handler
           progresses.

        Usually it is simpler and clearer in situations like the above to make
        the event handler for B call a helper function that would do what A was
        wanting to do.

        Of course, if the situation is very involved or has multiple handlers
        the flow might get too complex. We haven't seen this 'in the wild' often
        though.
```

Reply:
Awesome.

Reply:
Some possible alternative text. I'm not sure that I prefer my example to Chipaca's, so feel free to take the best of whatever text you like:
```
        Deferring an event from a handler puts that handler into a queue,
        to be called again the next time the charm is invoked. This may be an
        action or any event other than metric events. The queue of events will
        be dispatched before the new event is processed.

        From the above you may deduce, but it's important to point out:
        * ``defer()`` does not interrupt the execution of the current event
          handler. In almost all cases, a call to ``defer()`` should be followed
          by an explicit ``return`` from the handler;
        * the re-execution of the deferred event handler starts from the top of
          the handler method (not where defer was called), and
        * any deferred events get processed before the event (or action) that
          caused the current invocation of the charm.

        The general desire to call ``defer()`` happens when some precondition
        isn't yet met. However, care should be exercised if it is better to
        defer this event so that you see it again, or whether it is better to
        just wait for the event that indicates the precondition has been met.

        For example, if ``config-changed`` is fired, and you are waiting for
        different config. There is no reason to defer the event, because there
        will be a *different* ``config-changed`` event when the config actually
        changes, rather than checking to see if maybe config has changed prior
        to every other event that occurs.

        Similarly, if you need 2 events to occur before you are ready to
        proceed (say event A and B). When you see event A, you could chose to
        ``defer()`` it because you haven't seen B yet. However, that leads to:

        1. event A fires, calls defer()
        2. event B fires, event A handler is called first, still hasn't seen B
           happen, so is deferred again. Then B happens, which progresses since
           it has seen A.
        3. At some future time, event C happens, which also checks if A can
           proceed.
```

Reply:
I find your text covers what I wanted to say more concisely and with better flow overall, and I can't find how to improve it. I'll make sure it renders correctly and push it up.

Reply:
tweaked it a little bit, in the end. WDYT.